### DISCUSSION TITLE
Using layers, how do I import a _variables.scss in UI layer that is accessible in parent app?

### DISCUSSION THREAD

**Discussion by JeremyQtweb** [nuxt/nuxt#24473] [Questions]
I've been searching around the interweb for a bit but have not found a solution for my issue.

I have a parent app that extends a UI layer.
The UI layer has a ui_main.scss file and an additional _ui_variables.scss file that needs importing.

My parent app can access classes in the UI layers ui_main.scss file which is great.
In my UI layer however, it's ui_main.scss file cannot access the _ui_variables.scss file to use variables in there.

I know that CSS import is not the best in layers right now, as you have to use relative pathing and/or build the full path etc. to access files, which I've tried and **_does work_** for compiling the UI layers ui_main.scss file.
When I log the full path to the _ui_variables.scss file it **_IS_** correct too.

None of the obvious ways seem to allow my UI layer to 'see' it's variables file though, I've commented a couple of them in the code below.

NOTE: The UI layer does have Vuetify configured, not sure if it's `transformAssetUrls` function is affecting this.

**Parent nuxt.config**
```javascript
export default defineNuxtConfig({
	extends: ["../ProUiLayer", "../ProAuthLayer"],
	devtools: { enabled: true },
	css: ["~/assets/scss/main.scss"],
	vite: {
		css: {
			preprocessorOptions: {
				scss: {
					additionalData: '@import "~/assets/scss/_variables.scss";',
				},
			},
		},
	},
});
```

**Child nuxt.config (ProUiLayer)**
```javascript
import vuetify, { transformAssetUrls } from "vite-plugin-vuetify";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const currentDir = dirname(fileURLToPath(import.meta.url));
console.log(currentDir);
console.log(join(currentDir, "/assets/scss/_ui_variables.scss"));

export default defineNuxtConfig({
	devtools: { enabled: true },
	css: [join(currentDir, "./assets/scss/ui_main.scss")], // works fine, accessible in parent app
	build: {
		transpile: ["vuetify"],
	},
	modules: [
		(_options, nuxt) => {
			nuxt.hooks.hook("vite:extendConfig", (config) => {
				// @ts-expect-error
				config.plugins.push(vuetify({ autoImport: true }));
			});
		},
		//...
	],
	vite: {
		css: {
			preprocessorOptions: {
				scss: {
					additionalData: `@import "${join(currentDir, "./assets/scss/_ui_variables.scss")}";`, // doesn't work
					//additionalData: '@import "./assets/scss/_ui_variables.scss";', // doesn't work
					//includePaths: [join(currentDir, "./assets/scss")], // doesn't work
				},
			},
		},
		vue: {
			template: {
				transformAssetUrls,
			},
		},
	},
});
```

**This is the error I get...**
![image](https://github.com/nuxt/nuxt/assets/39040329/8f7fcda8-eeb6-4ea2-b19d-8584815f6918)

I guess a follow up question would be... how does this work if these layers are made into a NPM package? since these paths are local currently. Hmmm.

Any thoughts or suggestions would be most appreciated. I'll keep trying LOL

**Comment by JeremyQtweb** [comment]
Hey @danielroe - cheeky call for visibility :) Keen to figure this out as Nuxt Layers is awesome and just makes so much sense <3

**Comment by dsvgl** [comment]
@JeremyQtweb 
You set additionalData in your ui layer correctly. But then you effectively overwrite it in parent nuxt.config.
There is no need to overwrite/redefine it. It "transfers" automatically when extending the layer. Defining additionalData inside layer is enough.
See working example here: https://stackblitz.com/edit/github-9meqhs?file=.playground%2Fcomponents%2FHelloWorld.vue

If you want to "add" some vars to additionalData in the parent, you need to check if there is already some data and then append new stuff to it. But not sure how to check in parent nuxt.config.
I always use a nuxt module for this. There you can check for additionalData set the consuming nuxt.config and then append/prepend your data to it. Like so (nuxt2):
![SCR-20231207-jivm](https://github.com/nuxt/nuxt/assets/97700093/fa4bf35c-1dae-47f5-81e5-2a9b22a73b40)

I have not tested it together with Vuetifys transformAssetUrls. Try to add it to see if it is the culprit.

**Reply by dsvgl** [reply]
@JeremyQtweb does it help?

**Reply by JeremyQtweb** [reply]
@dsvgl - Sorry, only just getting back to this project now... umm... well it's given me the clue I needed, now to figure out how to resolve it.

I see the problem for sure. While I expected the UI layer's variable.scss file to 'transfer' like you say, I guess I also expected the additionalData property to be merged so both files would be defined at build time. It appears that I need to do that myself as you suggest.

What do you use a 'nuxt module' for exactly?, I don't understand what the implementation would look like for this purpose.
I'll probably recognise it when I see it, but right now I cannot visualise it in my head :P

**Reply by dsvgl** [reply]
@JeremyQtweb 
we have a monorepo with 10 nuxt apps and all share the same scss/css-variables (colors, margins, ...). We therefor implemented a Design System to have a single source of truth. We have a set of base variables (Design Tokens) and folders with overrides for each brand. All json files. We use [Style Dictionary](https://amzn.github.io/style-dictionary/#/) to output desired css and scss files.
The whole thing is done as a nuxt module that every app is using. We simply get the brands ID via publicRuntimeConfig and pass this as an option to our Style Dictionary script. This then knows in which folders it has to go to grab the correct jsons.
The output then is a brand specific css and scss file (with brandId in the name). Every brands css/scss file goes into a general (gitignored) folder in the monorepo.
Check the `scssTokensImportStatement` in the snippet below -> starting an app will automatically make it use its tailored css file as well as having access to the scss Design Tokens everywhere in the app. The value of these Tokens then can differ, eg. `$primary-color` being blue by default and red for app nr. 5... (BTW. the css is only for css custom properties).
See also the final line where the additionalData is set. The assignment is done via `+=`. So, if one app has its own additionalData definition, we would append the Design System path to it instead of overwriting it.
Hope this helps :)



```ts
// nuxt 2(!) module
import type { Module } from '@nuxt/types';
import { resolve } from 'path';
import type { DS_BrandId } from './tokens.config';
import { generateTokens } from './tokens.config';
import { TOKENS } from './utils';

const designSystemModule: Module = function () {
  const brandId: DS_BrandId = this.options.publicRuntimeConfig.BRAND_ID;

  const stylesDir: string = resolve(__dirname, '../../core/src/styles');
  const TOKENS_OUTPUT_DIR: string = `${stylesDir}/tokens`;
  const destinationFilename: string = `${TOKENS.namespace}.${brandId}`;

  this.nuxt.hook('build:before', () => {
    // trigger generation of Token css and scss files via Style Dictionary
    generateTokens(brandId);

    // NOTE: options.css is for "actual" styles!
    this.options.css.push(`${TOKENS_OUTPUT_DIR}/${destinationFilename}.css`);

    const scssTokensImportStatement: string = `@import '${TOKENS_OUTPUT_DIR}/_${destinationFilename}.scss';`;

    const additionalData: string[] = [
      // other stuff,
      scssTokensImportStatement,
    ];

    // NOTE: scss.additionalData is for variables only - no styles!
    this.options.build.loaders = this.options.build.loaders || {};
    this.options.build.loaders.scss = this.options.build.loaders.scss || {};
    this.options.build.loaders.scss.additionalData = this.options.build.loaders.scss.additionalData || '';
    this.options.build.loaders.scss.additionalData += additionalData.join(' ').trim();
  });
};

export default designSystemModule;
```
